main.c

```diff
@@ -36,98 +36,54 @@ struct {
-	char *argv[32], *arg, *argend;
+	int res;
+	unsigned int argc;
+
+	syspage_prog_t *prog;
+	char *argv[32], *cmdline;
 
 	/* Enable locking and multithreading related mechanisms */
 	_hal_start();
 
-	lib_printf("main: Decoding programs from data segment\n");
-	programs_decode(&main_common.kmap, &main_common.kernel);
-
 	lib_printf("main: Starting syspage programs:");
-	for (i = 0; i < syspage->progssz; i++)
-		lib_printf(" '%s',", syspage->progs[i].cmdline);
-	lib_printf("\b \n");
+	syspage_progShow();
 
 	posix_init();
 	posix_clone(-1);
 
-	/* Free memory used by initial stack */
-	/*vm_munmap(&main_common.kmap, main_common.stack, main_common.stacksz);
-	vm_pageFree(p);*/
-
-	/* Set stdin, stdout, stderr ports */
-//	proc_fileAdd(&h, &oid, 0);
-//	proc_fileAdd(&h, &oid, 0);
-//	proc_fileAdd(&h, &oid, 0);
-
-	argv[0] = NULL;
-
-	while (cmdline != NULL && *cmdline != '\0') {
-		end = cmdline;
-		while (*end && *(++end) != ' ');
-		while (*end == ' ')
-			*(end++) = 0;
-		if (*cmdline == 'X' && ++xcount) {
-			i = 0;
-			argend = cmdline;
-
-			while (i < sizeof(argv) / sizeof(*argv) - 1) {
-				arg = ++argend;
-				while (*argend && *argend != ';')
-					argend++;
+	/* Start programs from syspage */
+	if ((prog = syspage_progList()) != NULL) {
+		do {
+			cmdline = prog->argv;
+			/* If app shouldn't be executed then args should be discarded */
+			if (*cmdline != 'X') {
+				while (*cmdline != ';' && *cmdline != '\0')
+					++cmdline;
+
+				*cmdline = '\0';
+				continue;
+			}
 
-				argv[i++] = arg;
+			/* 'X' is no longer useful */
+			++prog->argv;
+			cmdline = prog->argv;
+			for (argc = 0; argc < (sizeof(argv) / sizeof(*argv) - 1);) {
+				argv[argc++] = cmdline;
+				while (*cmdline != ';' && *cmdline != '\0')
+					++cmdline;
 
-				if (!*argend)
+				if (*cmdline == '\0')
 					break;
 
-				*argend = 0;
+				*(cmdline++) = '\0';
 			}
-			argv[i++] = NULL;
-
-			if (i == sizeof(argv) / sizeof(*argv))
-				lib_printf("main: truncated arguments for command '%s'\n", argv[0]);
-
-			/* Start program loaded into memory */
-			for (i = 0; i < syspage->progssz; i++) {
-				if (!hal_strcmp(cmdline + 1, syspage->progs[i].cmdline)) {
-					argv[0] = syspage->progs[i].cmdline;
-					hal_memcpy(&prog, &syspage->progs[i], sizeof(prog));
-					res = proc_syspageSpawn(&prog, vm_getSharedMap(&prog, -1), prog.cmdline, argv);
-					if (res < 0)
-						lib_printf("main: failed to spawn %s (%d)\n", argv[0], res);
-				}
-			}
-		}
-
-		cmdline = end;
-	}
+			argv[argc++] = NULL;
 
-	if (!xcount && syspage->progssz != 0) {
-		argv[1] = NULL;
-		/* Start all syspage programs */
-		for (i = 0; i < syspage->progssz; i++) {
-			argv[0] = syspage->progs[i].cmdline;
-			hal_memcpy(&prog, &syspage->progs[i], sizeof(prog));
-			res = proc_syspageSpawn(&prog, vm_getSharedMap(&prog, -1), prog.cmdline, argv);
-			if (res < 0) {
+			if ((res = proc_syspageSpawn((syspage_prog_t *)prog, vm_getSharedMap(prog, -1), argv[0], argv)) < 0)
```

Comment:
```suggestion
			if ((res = proc_syspageSpawn(prog, vm_getSharedMap(prog, -1), argv[0], argv)) < 0)
```
`prog` is probably of the `syspage_prog_t*` type.

Reply:
Additional casting is not necessary, removed